File: HFSGladeGerador_java/src/hfsgladegerador/recursos/cpp/Rotinas.cpp
#include "resource.h"
#include "Rotinas.h"

#include <sql.h>
#include <sqlext.h>
// ---------------------------------------------------------------------------
CONEXAO dbConexao;
SQLHENV henv;
SQLHDBC hdbc;
// ---------------------------------------------------------------------------
TRotinas *Rotinas = (TRotinas *) NULL;
// ---------------------------------------------------------------------------
TRotinas::TRotinas() { 

}
// ---------------------------------------------------------------------------
TRotinas::~TRotinas() { 

}
// ---------------------------------------------------------------------------
void TRotinas::MsgDlg(Glib::ustring sMensagem, Gtk::MessageType tipo) {

  Gtk::MessageDialog dialog(retUTF8(sMensagem), false,
   tipo, Gtk::BUTTONS_OK, true);
  dialog.set_position(Gtk::WIN_POS_CENTER);
  dialog.set_has_separator(false);
  dialog.run();
}
// ---------------------------------------------------------------------------
Glib::ustring TRotinas::retLocale(Glib::ustring sTexto) {
  return Glib::locale_from_utf8(sTexto);
}
// ---------------------------------------------------------------------------
Glib::ustring TRotinas::retUTF8(Glib::ustring sTexto) {
  //return Glib::locale_to_utf8(sTexto);
  return sTexto;
}
// ---------------------------------------------------------------------------
bool TRotinas::ConectaBanco(Glib::ustring sFonteDeDados,
 Glib::ustring sLogin, Glib::ustring sSenha) {
   SQLRETURN rc;

   dbConexao = SQL_NULL_HSTMT;
   henv = SQL_NULL_HENV;
   hdbc = SQL_NULL_HDBC;

   // Aloca o ambiente ODBC e salva o manuseador.
   rc = SQLAllocHandle(SQL_HANDLE_ENV, SQL_NULL_HANDLE, & henv);

   if (rc == SQL_SUCCESS || rc == SQL_SUCCESS_WITH_INFO) {
     // Faz o ODBC saber que esta é uma aplicação ODBC 3.0.
     rc = SQLSetEnvAttr(henv, SQL_ATTR_ODBC_VERSION,
      (SQLPOINTER)SQL_OV_ODBC3, 0);

     if (rc == SQL_SUCCESS || rc == SQL_SUCCESS_WITH_INFO) {
       /* Aloca o manuseador da conexão */
       rc = SQLAllocHandle(SQL_HANDLE_DBC, henv, & hdbc);

       if (rc == SQL_SUCCESS || rc == SQL_SUCCESS_WITH_INFO) {
         /* Atribui login timeout para 5 segundos. */
         SQLSetConnectAttr(hdbc, SQL_LOGIN_TIMEOUT, (SQLPOINTER)5, 0);

         /* Conecta na fonte de dados */
         rc = SQLConnect(hdbc, (SQLTCHAR *) sFonteDeDados.c_str(), SQL_NTS,
          (SQLTCHAR *) sLogin.c_str(), SQL_NTS, (SQLTCHAR *) sSenha.c_str(), SQL_NTS);

         if (rc == SQL_SUCCESS || rc == SQL_SUCCESS_WITH_INFO) {
           /* Aloca manuseador do procedimento */
           rc = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, & dbConexao);

           if (rc == SQL_SUCCESS || rc == SQL_SUCCESS_WITH_INFO) {
             return true;
           } else
             SQLFreeHandle(SQL_HANDLE_STMT, dbConexao);
         } else
           SQLDisconnect(hdbc);
       } else
         SQLFreeHandle(SQL_HANDLE_DBC, hdbc);
     }
   } else
     SQLFreeHandle(SQL_HANDLE_ENV, henv);

   return false;

}
// ---------------------------------------------------------------------------
void TRotinas::DisconectaBanco() {
  SQLFreeHandle(SQL_HANDLE_STMT, dbConexao);
  SQLDisconnect(hdbc);
  SQLFreeHandle(SQL_HANDLE_DBC, hdbc);
  SQLFreeHandle(SQL_HANDLE_ENV, henv);
}
// ---------------------------------------------------------------------------
bool TRotinas::ConsultaDados(CONEXAO hstmtlocal, Glib::ustring sSql) {
  SQLRETURN rc;

  rc = SQLPrepare(hstmtlocal, (SQLTCHAR *) sSql.c_str(), SQL_NTS);
  if (!(rc == SQL_SUCCESS || rc == SQL_SUCCESS_WITH_INFO)) {
    MsgDlg(STR_NCPREPSQL, Gtk::MESSAGE_ERROR);
    return false;
  }
  rc = SQLExecute(hstmtlocal);
  if (!(rc == SQL_SUCCESS || rc == SQL_SUCCESS_WITH_INFO)) {
    MsgDlg(STR_NCEXECSQL, Gtk::MESSAGE_ERROR);
    return false;
  }
  if (rc == SQL_SUCCESS || rc == SQL_SUCCESS_WITH_INFO) {
    rc = SQLFetch(hstmtlocal);
    if (rc == SQL_NO_DATA) {
      SQLCloseCursor(hstmtlocal);
      return false;
    } else {
      SQLCloseCursor(hstmtlocal);
      return true;
    }
  } else {
    SQLCloseCursor(hstmtlocal);
    return false;
  }
}
// ---------------------------------------------------------------------------
bool TRotinas::AtualizaDados(CONEXAO hstmtlocal, Glib::ustring sSql) {
  SQLRETURN rc;

  rc = SQLPrepare(hstmtlocal, (SQLTCHAR *) retLocale(sSql).c_str(), SQL_NTS);
  if (rc == SQL_SUCCESS || rc == SQL_SUCCESS_WITH_INFO) {
    rc = SQLExecute(hstmtlocal);
    if (rc == SQL_SUCCESS || rc == SQL_SUCCESS_WITH_INFO)
      return true;
    else {
      return false;
    }
  } else {
    MsgDlg(STR_NCPREPSQL, Gtk::MESSAGE_ERROR);
    return false;
  }
}

<MetodosObjetos>

File: HFSGladeGerador_java/src/hfsgladegerador/recursos/php/Rotinas.php
<?php

define('STR_NCONBANCO','Não consigo conectar com o banco de dados!');
define('STR_NALOCODBC','Não foi possível alocar um manuseador para o ambiente ODBC');
define('STR_NCPREPSQL','Não conseguiu preparar o SQL!');
define('STR_NCEXECSQL','Não conseguiu executar o SQL!');	

class Rotinas
{
	private $dbConexao;

	function __construct()
	{
		//parent::__construct();			
	}
	
	function MsgDlg($sMensagem, $tipo) {
		$dialog = new GtkMessageDialog(null, Gtk::DIALOG_MODAL, $tipo, Gtk::BUTTONS_OK, $sMensagem);	
		$dialog->set_position(Gtk::WIN_POS_CENTER);
		$dialog->set_has_separator(false);
		$dialog->run();
		$dialog->destroy();
	}
	
	function ConectaBanco($sFonteDeDados, $sLogin, $sSenha) {
		try {
			$this->dbConexao = new PDO($sFonteDeDados);
			return true;
		} catch (PDOException $ex){
			print "ERRO [Rotinas.ConectaBanco]: " . $ex->getMessage();
			return false;
		}
	}

	function DisconectaBanco() {
		$this->dbConexao = null;
	}

	function ConsultaDados($sSql){		
		try {
			$stmt = $this->dbConexao->query($sSql);

			if($stmt) {
				/*
				foreach($stmt as $linha){
					print $linha['codigo']
				}				
				while ($linha = $stmt->fetch(PDO::FETCH_OBJ)){
					print $linha->codigo;
				}				
				*/
				if ($stmt->fetchObject()) 
					return true;
				else
					return false;	
					
			} else {
				return false;	
			}
		} catch (PDOException $ex){
			print "ERRO [Rotinas.ConsultaDados]: " . $ex->getMessage();
			return false;
		}
	}

	function AtualizaDados($sSql){		
		try {
			return $this->dbConexao->exec($sSql);
		} catch (PDOException $ex){
			print "ERRO [Rotinas.AtualizaDados]: " . $ex->getMessage();
			return null;
		}
	}

	<MetodosObjetos>
}

?>
